### PR TITLE
Introduce Stateful Plans

### PR DESCRIPTION
**Pull request by LaRiffle**
Response implementation to #2385

I made several choices in terms of API which can be summarized by these examples, they are of course very open to discussion if smthg more elegant can be found.

**Function 2 Plan**
```python
    @sy.func2plan(args_shape=[(1,)], state={"bias": th.tensor([1.0])})
    def foo(x, state):
        bias = state.read("bias")
        x = x * 2
        return x + bias
```
**Class 2 Plan**
```python
    class Net(sy.Plan):
        def __init__(self):
            super(Net, self).__init__()
            self.fc1 = nn.Linear(2, 1)
            self.bias = th.tensor([1000.0])

            self.add_to_state("fc1", "bias")

        def forward(self, x):
            x = self.fc1(x)
            return F.log_softmax(x, dim=0) + self.bias
```

**Known weaknesses**
- sy.Plan inherits from torch.nn.Module
- State elements are sent separately from the Plan (multiple msgs only when sending, not at execution of course)
- The code remains really hard to read

### CODE REVIEW

File: syft/messaging/plan.py [hunk 1]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+def method2plan(*args, **kwargs):
+    raise SyntaxError(
+        "method2plan is not supported anymore. Consider instead subclassing your object from sy.Plan."
     )
 
-    @property
-    def method(self: object) -> Plan:
+
+class State(object):
+    """The State is a Plan attribute and is used to send tensors along functions
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: Missing period.

File: syft/messaging/plan.py [hunk 2]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+
+    It references Plan tensor or parameters attributes using their name, and make
+    sure they are provided to remote workers who are sent the Plan.
+    """
+
+    def __init__(self, plan):
+        self.keys = []
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: Could self.keys be a set?

File: syft/messaging/plan.py [hunk 3]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+        self.plan = plan
+
+    def __repr__(self):
+        return "State: " + ", ".join(self.keys)
+
+    def append(self, key):
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: Can you add more details to this docstring? There's a lot going on here, not sure if I follow.

File: syft/messaging/plan.py [hunk 4]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+                raise ValueError(
+                    f"Obj of type {type(t)} is not supported in the state.\n"
+                    "Use instead tensors or parameters."
+                )
+        return self
+
+    def __iadd__(self, other_keys):
+        return self.append(other_keys)
+
+    def is_accepted_type(self, obj):
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: Maybe rename this to `is_valid_type`?

File: syft/messaging/plan.py [hunk 5]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+
+    def read(self, key):
+        """
+        Returns one element referenced by the store
+        """
+        return getattr(self.plan, key)
+
+    def set_(self, dict_state):
+        """
+        Reset inplace the state by feeding him a dict of tensors or params
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: him -> it

File: syft/messaging/plan.py [hunk 6]
```diff
@@ -33,62 +33,134 @@ class func2plan(object):
...
+class State(object):
+    """The State is a Plan attribute and is used to send tensors along functions
+
+    It references Plan tensor or parameters attributes using their name, and make
+    sure they are provided to remote workers who are sent the Plan.
+    """
+
+    def __init__(self, plan):
+        self.keys = []
+        self.plan = plan
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: The state is a plan attribute, but a state has a plan attribute, can you detail why this is the case?

File: syft/messaging/plan.py [hunk 7]
```diff
@@ -99,30 +171,38 @@ class Plan(ObjectStorage):
...
+        self.name = self.__class__.__name__ if name == "" else name
+        self.owner = sy.local_worker if owner is None else owner
         self.verbose = verbose
 
         # Info about the plan stored
         self.plan = list()
         self.readable_plan = readable_plan if readable_plan is not None else []
+        self.state = State(self)
+        self.include_state = include_state
+        if state is not None:
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: A state can store another state?

File: syft/messaging/plan.py [hunk 8]
```diff
@@ -134,16 +214,20 @@ def __init__(
...
+
+    def add_to_state(self, *elements):
+        for elem in elements:
+            self.state.append(elem)
 
     def _auto_build(self, args_shape: List[Tuple[int]] = None):
         placeholders = self._create_placeholders(args_shape)
-        args = [self._self] + placeholders if self.is_method else placeholders
-        self._build(args)
+        args = placeholders
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: I think you can get rid of this variable and just call placeholders args instead

File: syft/messaging/plan.py [hunk 9]
```diff
@@ -190,59 +288,53 @@ def _recv_msg(self, bin_message: bin):
 
         return sy.serde.serialize(None)
 
-    def _prepare_for_running_local_method(self):
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: I'm very happy to see this is gone :smiley:

File: syft/frameworks/torch/hook/hook.py [hunk 10]
```diff
@@ -546,6 +546,7 @@ def create_grad_objects(model):
                 o = p.sum()
                 o.backward()
                 p.grad -= p.grad
+            pass
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/hook/hook.py]: nope I thought it was removed!

File: syft/frameworks/torch/hook/hook.py [hunk 11]
```diff
@@ -646,24 +649,24 @@ def module_copy(nn_self):
...
+        # def location(nn_self):
+        #     try:
+        #         for p in nn_self.parameters():
+        #             return p.location
+        #     except AttributeError:
+        #         raise AttributeError(
+        #             "Module has no attribute location, did you already send it to some location?"
+        #         )
+        #
+        # self.torch.nn.Module.location = location
```
[the remark this answers is not in the document]
Comment (LaRiffle) [reply syft/frameworks/torch/hook/hook.py]: Will put it back!

File: syft/messaging/plan.py (hunk 2 above)
Comment (LaRiffle) [reply syft/messaging/plan.py]: totally

File: syft/messaging/plan.py (hunk 6 above)
Comment (LaRiffle) [reply syft/messaging/plan.py]: The state should theoretically be in the plan class, but I moved it outside for clarity. But the state does not store the elements, juste references to elements, which are in thee plan. That's why the both reference each other.
Which is not optimal...

File: syft/messaging/plan.py (hunk 7 above)
Comment (LaRiffle) [reply syft/messaging/plan.py]: Not at all, but it can be fed with a dict like in self.state.set_(state)

File: test/message/test_plan.py [hunk 12]
```diff
@@ -127,15 +240,31 @@ def forward(self, x):
...
+            self.state += ["fc1", "bias"]
+
+        def forward(self, x):
+            x = self.fc1(x)
+            return F.log_softmax(x, dim=0) + self.bias
+
+    model = Net()
+
+    # Force build
+    assert model(th.tensor([1.0, 2])) == th.tensor([1000.0])
```
Comment (mari-linhares) [inline test/message/test_plan.py]: there's a build method implemented, I think we should use it explicitly.

Comment (mari-linhares) [reply test/message/test_plan.py]: Like you've done below with net.build(tensor)

File: test/message/test_plan.py [hunk 13]
```diff
@@ -226,30 +381,47 @@ def plan_abs(data):
...
 
-    hook.local_worker.is_client_worker = True
+    x_ptr = th.tensor([-1, 9, 3]).send(alice)
+    p = plan_abs(x_ptr)
+    x_abs = p.get()
+    assert (x_abs == th.tensor([2, 10, 4])).all()
+
+
+# TODO: Clarify this test
+# def test_fetch_plan(hook):
```
Comment (mari-linhares) [inline test/message/test_plan.py]: The idea of fetching a plan is to get a copy of an existing plan that I (the worker) can execute. We need this for providing secure serving for Grid (which I'm working right now), I'm happy to discuss this further on slack.

File: test/message/test_plan.py [hunk 14]
```diff
@@ -473,62 +643,6 @@ def test__replace_message_ids():
     assert replaced == expected
 
 
-def test___call__function(hook):
```
Comment (mari-linhares) [inline test/message/test_plan.py]: there's no way to reuse these tests?

File: syft/messaging/plan.py [hunk 15]
```diff
@@ -190,59 +311,53 @@ def _recv_msg(self, bin_message: bin):
...
+            res_ptr = self.forward(*local_args)
         res_ptr.child.garbage_collect_data = False
 
-        worker = self.find_location(args)
+        # We put back the original state. There is no need to keep a reference
+        # To the element sent to the plan.
+        self.state.set_(copied_state)
 
+        # The readable plan is now built, we hide the fact that it was run on
+        # the plan and on the owner by replacing the workers ids
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: the plan and on the owner by replacing the workers ids -> the plan and not on the owner by replacing the workers ids

File: test/message/test_plan.py (hunk 14 above)
Comment (LaRiffle) [reply test/message/test_plan.py]: I don't know really what to do with these tests, I can hardly them fit with the new notions of stateful plans :/

Comment (mari-linhares) [reply test/message/test_plan.py]: I see, that's okay, I guess. :disappointed:

File: syft/messaging/plan.py [hunk 16]
```diff
@@ -736,6 +736,9 @@ def __str__(self):
 
         return out
 
+    def __repr__(self):
```
Comment (mari-linhares) [inline syft/messaging/plan.py]: Nice

**Pull request merged**